tests/test_pp.py

```diff
@@ -0,0 +1,382 @@
+    """
+    adata, df_cov, df_gs, dic_res_ref = load_toy_data()
+    adata = adata[:, :50].copy()
+    scdrs.pp.preprocess(adata, cov=None)
+
+    assert "SCDRS_PARAM" in adata.uns, "'SCDRS_PARAM' should be in adata.uns"
+    KEY_LIST = [
+        "FLAG_SPARSE",
+        "FLAG_COV",
+        "GENE_STATS",
+        "CELL_STATS",
+    ]
+    for key_ in KEY_LIST:
+        assert key_ in adata.uns["SCDRS_PARAM"], (
+            "'%s' should be in adata.uns['SCDRS_PARAM']" % key_
+        )
+    assert adata.uns["SCDRS_PARAM"]["FLAG_SPARSE"] is True, "FLAG_SPARSE=True"
+    assert adata.uns["SCDRS_PARAM"]["FLAG_COV"] is False, "FLAG_COV=False"
+
+    return
+
+
+def test_preprocess_dense_cov():
+    """
+    Test scdrs.pp.preprocess: dense + cov (normal mode)
+    """
+    adata, df_cov, df_gs, dic_res_ref = load_toy_data()
+    adata = adata[:, :50].copy()
+    adata.X = adata.X.toarray()
+    scdrs.pp.preprocess(adata, cov=df_cov)
+
+    assert "SCDRS_PARAM" in adata.uns, "'SCDRS_PARAM' should be in adata.uns"
+    KEY_LIST = [
+        "FLAG_SPARSE",
+        "FLAG_COV",
+        "GENE_STATS",
+        "CELL_STATS",
+    ]
+    for key_ in KEY_LIST:
+        assert key_ in adata.uns["SCDRS_PARAM"], (
+            "'%s' should be in adata.uns['SCDRS_PARAM']" % key_
+        )
+    assert adata.uns["SCDRS_PARAM"]["FLAG_SPARSE"] is False, "FLAG_SPARSE=False"
+    assert adata.uns["SCDRS_PARAM"]["FLAG_COV"] is True, "FLAG_COV=True"
+
+    return
+
+
+def test_preprocess_dense_nocov():
+    """
+    Test scdrs.pp.preprocess: dense + nocov (normal mode)
+    """
+    adata, df_cov, df_gs, dic_res_ref = load_toy_data()
+    adata = adata[:, :50].copy()
+    adata.X = adata.X.toarray()
+    scdrs.pp.preprocess(adata, cov=None)
+
+    assert "SCDRS_PARAM" in adata.uns, "'SCDRS_PARAM' should be in adata.uns"
+    KEY_LIST = [
+        "FLAG_SPARSE",
+        "FLAG_COV",
+        "GENE_STATS",
+        "CELL_STATS",
+    ]
+    for key_ in KEY_LIST:
+        assert key_ in adata.uns["SCDRS_PARAM"], (
+            "'%s' should be in adata.uns['SCDRS_PARAM']" % key_
+        )
+    assert adata.uns["SCDRS_PARAM"]["FLAG_SPARSE"] is False, "FLAG_SPARSE=False"
+    assert adata.uns["SCDRS_PARAM"]["FLAG_COV"] is False, "FLAG_COV=False"
+
+    return
+
+
+def test_preprocess_consistency():
+    """
+    Test scdrs.pp.preprocess: consistenty between sparse+cov and dense+cov
+    """
+    adata, df_cov, df_gs, dic_res_ref = load_toy_data()
+    adata_sparse = adata[:, :50].copy()
+    scdrs.pp.preprocess(adata_sparse, cov=df_cov)
+
+    adata_dense = adata[:, :50].copy()
+    adata_dense.X = adata_dense.X.toarray()
+    scdrs.pp.preprocess(adata_dense, cov=df_cov)
+
+    # COV
+    mat_X_sparse = (
+        adata_sparse.X.toarray()
+        + adata_sparse.uns["SCDRS_PARAM"]["COV_MAT"].values.dot(
+            adata_sparse.uns["SCDRS_PARAM"]["COV_BETA"].values.T
+        )
+        + adata_sparse.uns["SCDRS_PARAM"]["COV_GENE_MEAN"].values
+    )
```

Comment:
since this sub-routine has been used repeatedly and not so straightforward.

it may be worthwhile to add a function like scdrs.method.sparse_to_dense(adata_sparse) -> X_dense

where it checks the presence of adata_sparse.uns["SCDRS_PARAM"] and return the X_dense

**Edit:** I realized such function will never used by user and will only be used as testing purpose. So it may be meaningless to have such wrapper.

Reply:
Yeah probably no need for a wrapper.